### DISCUSSION TITLE
Proposal to Implement Service Binding Support

### DISCUSSION THREAD

**Discussion by bennesp** [cloudnative-pg/cloudnative-pg#2370] [Ideas]
[Service Binding](https://servicebinding.io/) is an open-source project that aims to standardize the exposure and binding of backing service secrets to applications. It provides a straightforward and consistent method to connect applications with services such as databases, essentially doing away with the need for manual configuration of secrets, config maps, and other resources.

Implementing Service Binding in the Cloud Native Postgres Operator would offer several benefits:

1. Ease of Platform Creation (to me this is the most important benefit): For platform engineers, Service Binding simplifies the process of creating platforms for developers. It provides a consistent and secure method of connecting services to applications.
2. Standardization: As an open-source project, Service Binding is proposing as a standard for service-to-application connections in the Kubernetes ecosystem. Adopting it would align the Cloud Native Postgres Operator with other operators that already support Service Binding.
3. Simplified Service Connection: Service Binding reduces the complexity of connecting applications to services, making it easier for developers to focus on their application logic rather than service management.

Some operators already are integrated with Service Binding.
Examples given [here](https://servicebinding.io/service-provider/#real-world-examples):
- [RabbitMQ Operator](https://github.com/rabbitmq/cluster-operator/pull/615)
- [Kafka Access Operator](https://github.com/strimzi/kafka-access-operator)
- [External Secrets Operator](https://external-secrets.io/)

According to [the specification](https://github.com/servicebinding/spec#provisioned-service), supporting Service Binding requires cloudnative-pg operator to define a `.status.binding` field in the database, which is reference to a Secret in the same namespace of the database. This should be quite easy to implement, of course if the proposal could be something that maintainers would like.

**Comment by phisco** [comment]
Could you elaborate on the expected content of the secret? It looks like there are some well-known fields, but the specification is allowing additional ones

**Comment by phisco** [comment]
Just to make it clear, you are talking about the application secret only (the one with the `-app` suffix), right?

**Comment by bennesp** [comment]
Currently, for a database called `my-db` with user `app` these Secrets are created (correct me if I'm wrong):
-  `my-db-ca`: an `Opaque` Secret, containing `ca.crt` and `ca.key`
- `my-db-app` and `my-db-superuser`: two `kubernetes.io/basic-auth` Secrets, containing `username`, `password` and `pgpass`
- `my-db-replication` and `my-db-server`, two `kubernetes.io/tls` Secrets, containing `tls.crt` and `tls.key`

In the Service Binding spec, under [Well Known Secret Entries](https://github.com/servicebinding/spec#well-known-secret-entries), there are some interesting hints. There they suggest some fields: `username`, `password`, `host`, `port`, `uri`, `certificates` and `private-key`.

I am not familiar with how cloudnative-pg expects users to connect with mTLS (sorry for that, I am still learning!) but I expect to have, in the end, all the pieces of information I need to securely connect to an instance of the database.

So, for sure I will need `username` and `password`, which are currently in the `-app` suffix Secret, but probably I, as a developer needing a database, would also need a `uri` and probably some tls certificates/keys for mTLS (?)

Also, we could think of starting supporting Service Binding with just the `username` and `password` credentials and extend it afterwards

**Comment by sxd** [comment]
Hi @bennesp 
You have all the information you know
You have:
```
Cluster name
Username
Password
Port
Namespace
```
What is missing here? it's this something that will be use to connect external service to apps inside kubernetes or is intend to be used inside kubernetes  like an app inside kubernetes? 

From the page you pointed I can read this:
```
Today in Kubernetes, the exposure of secrets for connecting application workloads to external services such as REST APIs, databases, event buses, and many more is manual and bespoke.
```
This is not meant to be used for services running inside kubernetes but for external service rather than something else, in this case you have everything to connect.

Related to the ow to use mTLS you can read here https://cloudnative-pg.io/documentation/1.20/ssl_connections/#issuing-a-new-certificate

I'm turning this into a discussion since it's not an issue on the operator and deserves some discussion.

Best Regards!

**Comment by sxd** [comment]
@bennesp BTW I'm not totally against more fields, it's just adding fields that will make sense and not over information.

**Comment by bennesp** [comment]
Thank you for taking time to read this @sxd.

## External services

To me, the boundary between an "external service" and a non-external one is somehow blurry, even more in the context of Service Binding.

Other operators (like the RabbitMQ one for example) are indeed operators to create resources inside the Kubernetes cluster, but they could be seen as "external services" from the developers point of view. Let me explain what I mean.

Today we live in a cloud world where Internal Developer Platforms are a thing, where a new job position called Platform Engineer exist. I am sure you are aware of this, but for everyone reading, very briefly, the job of a Platform Engineer is to allow developers to develop abstracting the Kubernetes part away from them, so that they can focus on writing code.

To me, Service Binding is just a very simple but effective tool to connect Pods with services without the need of knowing anything about Kubernetes: without knowing which Service to use, which Secrets to use, without even accessing Secrets. This is not always desirable of course, and that's why I don't want at all to leave such an abstraction to cloudnative-pg itself: cloudnative-pg users **must** be able to choose every single part, from which Service to use to which Secret and so on. But Service Binding could give us an interesting way of abstracting connection details.

## Scenario without Service Binding

This is how I imagine a scenario with the current state-of-the-art, so without Service Binding:
- Developers declare that they want a database with a simple declaration (that could be the cnpg `Cluster` or even a simpler one or even from a UI, that's not important, but at the very end a controller will create a `Cluster`)
- Developers want to connect to the database
- Platform Engineers must communicate the **convention** used to developers so that developers can mount the correct Secret**s** (either manually or automatically) and they can get the knowledge of which Service to connect to.
- Given that, developers now know how to get all the connection details and they can connect to the database.

## Scenario with Service Binding

In this case though:
- (the same) Developers declare that they want a database with a simple declaration (that could be the cnpg `Cluster` or even a simpler one or even from a UI, that's not important, but at the very end a controller will create a `Cluster`)
- (the same) Developers want to connect to the database
- Instead of Platform Engineers having to communicate the **convention** for all the connection details (either manually or in an automatic fashion), they just tell Developers to declare a ServiceBinding.
- Developers then create a ServiceBinding saying that they want to connect their backend deployment with the database they declared they wanted to be created. The only piece of information they need to know is the name of the database. 
- ServiceBinding then will be the only responsible to give to the backend deployment all the connection details, including username, password, uri and tls credentials in case they are needed.

## Conclusions

This is the reason why I was talking about more than just username and password.

Overall I don't think Service Binding is strictly necessary: Platform Engineers could just create the machinery to automatically mount Secrets and Service connection details in the needed Pod, but that is just mimicking what Service Binding is currently capable of doing in a "standard" way (instead of using a custom approach).

**Comment by bennesp** [comment]
Also, I would like to underline that a potential approach to this could just be to create another different `Secret`, leaving the `-app` one exactly as it is, while aggregating all the necessary information inside a new different Secret

**Comment by jcpunk** [comment]
I'd be fine either way.  I'm not sure how disruptive it would be to extend the `-app` secret.  Extra fields *probably* aren't too bad - unless folks are expecting those fields not to be there, but changing the type probably would have an impact out in the wild.

**Reply by bennesp** [reply]
> but changing the type probably would have an impact out in the wild.

Please notice that, per spec, this is a **SHOULD** and not a **MUST**, so we could leave the secret type as it is:

> The Secret type (.type verses .data.type fields) SHOULD reflect this value as servicebinding.io/{type}, substituting {type} with the Secret data type

**Comment by bennesp** [comment]
@sxd or @phisco would you be interested in accepting a PR for this?

Still not sure if it could be better to have a "new" ad-hoc Secret with all the necessary info or to use the existing `-app` one, enriching it.

Thanks

**Comment by lion7** [comment]
It's been a while since there was activity here, but I would also love to see CNPG to support service binding.

In my use-case, I'm both the platform engineer and app developer. I'm currently creating a Java application that I'd like to bind to my CNPG database. Currently I'm passing the CNPG JDBC uri (containing the URL + credentials)  directly as an env variable like so:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: backend
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/name: backend
  template:
    metadata:
      labels:
        app.kubernetes.io/name: backend
    spec:
      containers:
        - name: backend
          image: my-backend-image
          imagePullPolicy: Always
          env:
            - name: SPRING_DATASOURCE_URL
              valueFrom:
                secretKeyRef:
                  name: my-db-app
                  key: jdbc-uri
...
```

Ideally, I'd like to replace this with a service binding. This is well supported by frameworks as is [shown here with Spring](https://zarinfam.medium.com/3717d3486886?sk=346dc534327888ca805aad94e5d0f1b5) and [explained here with Quarkus](https://quarkus.io/guides/deploying-to-kubernetes#introduction-to-the-service-binding-operator). 

TLDR: Would you be willing to accept a PR that add support to CNPG? Most of the work that has to be done is already explained above. Other operators such as `External Secrets` already have support, which should give enough guidance for the [implementation](https://github.com/external-secrets/external-secrets/blob/fa8a58a2cf5bfeed9833dd2fccf6464a3d3453f6/design/design-crd-spec.md?plain=1#L210) and [usage](https://github.com/servicebinding/runtime/tree/main/samples/external-secrets).
